### PR TITLE
auto generate view_* methods

### LINKED ISSUE #977: Test `view_*` API
## 🚀 Feature
As noted in #976 it might be nice to add a test that our `view_*` API looks like our `add_*` and `Viewer` APIs combined. This should be true for all layers. We should test both the function signature and doc strings. We can test that the keyword args etc are even in the same order.

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
Following conversation in #976 about keeping all of our `add_*`,`view_*`, and `Viewer.__init__` methods in sync, here's a crazy little bit of code that will autogenerate all of our `view_*` methods, while retaining their current function signatures, introspectability, tab completion in IPython, and sphinx autodoc.  It combines the `Viewer.__init__` and `Viewer.add_*` method signatures and docstring parameters.  I haven't yet been able to find any practical differences between this and the current code.  I tried to comment inline with explanations.

Let me know what you think...

This would close #977

## Type of change
- [x] New feature (non-breaking change which adds functionality)

# How has this been tested?

- [x] all tests pass with my change

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by sofroniewn — APPROVED** [review]
Wow @tlambert03 this is crazy! Any PR that removes > 600 lines of code for the same functionality seems pretty cool to me!! I am approving this, but I'll let @jni and @kne42 weigh in before merge.

Out of curiosity can you extend any of this approach to our `add_*` methods (in follow up PRs)? It might be trickier there now that we allow for some API divergence around the multichannel stuff in `add_image` and some of the `path` file loading code. Still curious what your take is?

**Comment by tlambert03** [comment]
It's conceptually possible to do it for the `add_*` methods, but much harder in some cases due to divergence between the `Layer.__init__` method and the corresponding `add_*` method.  For instance, in the `add_image` method we accept `lists` of values for a number of parameters like colormap, contrast_limits, gamma, etc...  So we can't just copy the docstrings and signature from `Image.__init__`.  As such, I think it's probably either too brittle or error prone to autogenerate, and/or won't actually be that much more DRY.

**Comment by sofroniewn** [comment]
> It's conceptually possible to do it for the add_* methods, but much harder in some cases due to divergence between the Layer.__init__ method and the corresponding add_* method.

Yup, makes sense. Useful to think about though, as it might be a nice goal to keep in mind for the future

**Comment by jni** [comment]
> I am approving this, but I'll let @jni and @kne42 weigh in before merge.

I am less gung-ho about this one. It removes a lot of code, which is very fancy and clever, but perhaps too clever. I certainly don't understand what's happening on a quick glance and would like for more than one of us to do so before merging. I also would like to try it out on pycharm to verify that it doesn't adversely affect introspection there.

Not saying at all I don't want this merge, just that it's important to make the code readable for when things go wrong.

**Comment by tlambert03** [comment]
Yeah I share @jni’s view here.  This one is kinda freaky.  I think I’m generally in a favor of it, since it will guarantee parity between those methods and (reduce human error) but it’s pretty unconventional, so let’s let everyone sit with it for a bit. 

In terms of readable code, I tried to comment pretty well, but can definitely be even more verbose if you’d like, or add more to the docstring notes

File: napari/view_layers.py
Comment (tlambert03) [inline napari/view_layers.py]: This line is really the meat of it.  Here's how it works:

1. we define `real_func`, which is the (easier to understand) function that will do the work of creating a new viewer and adding a layer to it.
2. we create a **string** (`fakefunc`) that represents how we _would_ have typed out the original `view_*` method. 
    - `{combo_sig}` is an [inspect.Signature](https://docs.python.org/3/library/inspect.html#inspect.Signature) object (whose string representation is, conveniently, exactly how we would have typed the original function).
    - the inner `real_func({inner_sig})` part is basically how we were typing the body of the `view_*` functions before, e.g.: `(data=data, name=name, scale=scale ...)`
3. we compile that string into `view_func_code`
4. finally, we actually evaluate the compiled code in a (safe) empty namespace, and provide a `locals()` dict that tells python that the function name `real_func` in the `fakefunc` string actually corresponds to the `real_func` that we defined on line 66.   (Note: evaluation at this step is essentially exactly what was previously happening when python hit each `def view_*` declaration when importing `view_layers.py`)
5. we grab our newly evaluated function from the empty namespace that we supplied... and go on to add our documentation below

**Review by kne42 — COMMENTED** [review]
awesome PR @tlambert03!

while I am a huge fan of autogenerating duplicated functions, I am :-1: on this for the same reason why we copy/pasted the layer signatures to the add_* methods: certain IDEs (*cough\* PyCharm *cough*) opt to statically parse documentation instead of dynamically fetching it for auto-complete and documentation which means that these functions would not show up for their users

**Comment by tlambert03** [comment]
> while I am a huge fan of autogenerating duplicated functions, I am 👎 on this for the same reason why we copy/pasted the layer signatures to the add_* methods: certain IDEs (*cough* PyCharm _cough_) opt to statically parse documentation instead of dynamically fetching it for auto-complete and documentation which means that these functions would not show up for their users

wow, that sucks!  the code autocompletion and intellisense work just fine with this code in vscode ... come on pycharm, that's quite frustrating.

**Comment by kne42** [comment]
@tlambert03 we can check again and pray that PyCharm updated this, as the last time this issue arose was ~1 year ago but yeah it's rly dumb :/

**Comment by kne42** [comment]
or maybe it has a plugin for jedi? lol

**Comment by tlambert03** [comment]
I just downloaded pycharm and can confirm that it still doesn't work. 😡. lame.

**Comment by tlambert03** [comment]
I’m going to close this, since pycharm usage is common enough that this would be unacceptable.  If either pycharm capabilities improve (or we care less about supporting it) in the future, we can revisit

**Comment by tlambert03** [comment]
reopening in case we consider not worrying about pycharm

**Comment by sofroniewn** [comment]
> reopening in case we consider not worrying about pycharm

I'm supportive of merging this. Maybe @jni can review

File: napari/view_layers.py
Comment (jni) [inline napari/view_layers.py]:
```suggestion
function signature is a merged version of one of the ``Viewer.__init__`` method
and the ``Viewer.add_<layer_type>`` methods.  The final generated functions
follow this pattern
```

File: napari/view_layers.py
Comment (jni) [inline napari/view_layers.py]: I'd remove the line number, that is just *asking* to be made instantly obsolete.

File: napari/view_layers.py
Comment (jni) [inline napari/view_layers.py]: Should we just make these keyword-only to begin with anyway?

File: napari/view_layers.py
Comment (jni) [inline napari/view_layers.py]:
```suggestion
view_kwargs = {
    k: kwargs.pop(k) for k in kwargs if k in viewer_sig.parameters
}
```

File: napari/view_layers.py
Comment (jni) [inline napari/view_layers.py]: Can you make this a separate function, `_combine_docstrings` for example?

**Review by jni — APPROVED** [review]
Great stuff! I'm super happy about this. We indeed chose not to do this because it confused PyCharm, but at this point I think that is PyCharm's problem rather than ours. Having the functions repeated everywhere has been frustrating when adding/changing keyword arguments, and this solves that problem, and is perfectly Pythonic.

FWIW the `add_` methods can also be autogenerated after this... =)

File: napari/view_layers.py
Comment (tlambert03) [reply napari/view_layers.py]: sure... add the star here?
```python
class Viewer:
    def __init__(
        self,
        *,
        title='napari',
        ndisplay=2,
        order=None,
        axis_labels=None,
        show=True,
    ):
```

File: napari/view_layers.py
Comment (tlambert03) [reply napari/view_layers.py]: hah.  _this_ broke my test :). `RuntimeError: dictionary changed size during iteration`. had to put the `kwargs` iterator in a list

File: napari/view_layers.py
Comment (jni) [reply napari/view_layers.py]: Yep!

File: napari/view_layers.py
Comment (jni) [inline napari/view_layers.py]: Rename to `_generate_view_function` or similar now, so that it's clear that this is doing more than just building?

File: napari/view_layers.py
Comment (tlambert03) [reply napari/view_layers.py]: done

**Comment by sofroniewn** [comment]
Ok, this looks all good to go, any PR that removes 700 lines of code is very exciting! I'll let sit a little more in case @jni wants to look on more time when he wakes up, but then I'll merge!!

**Pull request merged**
